Define setLoginStatus

pooneh-nb opened the pull request:
#### eef2171648310d393bc7f65593a452755e5b45d5
<pre>
Define setLoginStatus
<a href="https://bugs.webkit.org/show_bug.cgi?id=276346">https://bugs.webkit.org/show_bug.cgi?id=276346</a>

Reviewed by Charlie Wolfe.

Define setLoginStatus and add code to send the status to network process.
The status is kept in memory (in a hashmap) in WebResourceLoadStatisticsStore.
In the next step, we will keep this status persistently in a DB on disk.

* LayoutTests/http/tests/is-logged-in/available-in-secure-contexts.https.html:
* LayoutTests/http/tests/login-status/login-status-api-getter-setter-functions-expected.txt: Added.
* LayoutTests/http/tests/login-status/login-status-api-getter-setter-functions.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setLoginStatus):
(WebCore::ChromeClient::isLoggedIn):
* Source/WebCore/page/IsLoggedIn.h: Copied from Source/WebCore/page/Navigator+LoginStatus.idl.
* Source/WebCore/page/IsLoggedIn.idl: Added.
* Source/WebCore/page/Navigator+LoginStatus.idl:
* Source/WebCore/page/NavigatorLoginStatus.cpp:
(WebCore::NavigatorLoginStatus::setStatus):
(WebCore::NavigatorLoginStatus::hasSameOrigin const):
(WebCore::NavigatorLoginStatus::isLoggedIn):
(WebCore::NavigatorLoginStatus::setLoggedIn): Deleted.
(WebCore::NavigatorLoginStatus::setLoggedOut): Deleted.
* Source/WebCore/page/NavigatorLoginStatus.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setLoginStatus):
(WebKit::WebResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setLoginStatus):
(WebKit::NetworkConnectionToWebProcess::isLoggedIn):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setLoginStatus):
(WebKit::WebChromeClient::isLoggedIn):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLoginStatus):
(WebKit::WebPage::isLoggedIn):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/281308@main">https://commits.webkit.org/281308@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/070d1be36a0b28df9bd7dc32b35e28dcad8d0912

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38811 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63382 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9978 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61596 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46465 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10142 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7008 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61497 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51482 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29101 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/58994 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32955 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8733 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8914 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54906 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9014 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65114 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3395 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/55616 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3406 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51475 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55724 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13168 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2828 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34626 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/35709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36795 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/35454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/page/ChromeClient.h:
```diff
@@ -59,6 +59,7 @@
 #include "WebGPU.h"
 #include "WorkerClient.h"
 #include <JavaScriptCore/ConsoleTypes.h>
+#include <cstdint>
```
This include probably isn't needed. If clangd is adding these you can add `--header-insertion=never` to your language server arguments.

charliewolfe commented on Source/WebCore/page/LoginStatus.h:
```diff
@@ -42,7 +47,7 @@ class LoginStatus {
 
     enum class CredentialTokenType : bool { LegacyCookie, HTTPStateToken };
-    enum class AuthenticationType : uint8_t { WebAuthn, PasswordManager, Unmanaged };
+    enum class AuthenticationType : uint8_t { WebAuthn, PasswordManager, PassKey, Unmanaged };
```
We should remove this change, and add it back in a future PR.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -26,11 +26,22 @@
+#include "dom/Document.h"
+#include "dom/DocumentStorageAccess.h"
+#include <wtf/StdLibExtras.h>
+#include <wtf/text/WTFString.h>
```
Many of these includes are likely not needed.

pooneh-nb replied:
But I need Document to get access to frame and page, no?
```
Document* document = m_navigator.document();
RefPtr frame = document->frame();
 page = frame->page();
```

charliewolfe replied:
Right, you probably need document, but you can remove the `dom/` and change the include to just `#include "Document.h"`.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -26,11 +26,22 @@
+#include "dom/DocumentStorageAccess.h"
+#include <wtf/StdLibExtras.h>
+#include <wtf/text/WTFString.h>
+
```
This extra new line should be removed.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,32 +58,43 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
+    else if (status == WTF::String::fromLatin1("logged-out"))
+        m_loginStatus = IsLoggedIn::No;
     else
         promise->reject();
```
We should return from this function here, since we are rejecting the promise.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,32 +58,43 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
-void NavigatorLoginStatus::setLoggedOut(Ref<DeferredPromise>&& promise)
-{
-    promise->resolve();
+    Document* document = m_navigator.document();
```
We should use a smart pointer here for Document. We also probably need a null check.
```suggestion
    RefPtr document = m_navigator.document();
```

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.h:
```diff
@@ -26,31 +26,34 @@
 #pragma once
 
 #include "Supplementable.h"
+#include <wtf/WeakPtr.h>
```
This should be removed.

pooneh-nb replied:
I believe we need this as we are extending  public CanMakeWeakPtr<NavigatorLoginStatus>. Correct me if I'm wrong

charliewolfe replied:
Oh you're right.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.h:
```diff
@@ -26,31 +26,34 @@
 class DeferredPromise;
 class Navigator;
-class NavigatorLoginStatus final : public Supplement<Navigator> {
+class WeakPtrImplWithEventTargetData;
```
This should be removed.

pooneh-nb replied:
you mean remove the extending navigator?

charliewolfe replied:
The forward declaration should be removed (`class WeakPtrImplWithEventTargetData;`)

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.h:
```diff
@@ -26,31 +26,34 @@
     static ASCIILiteral supplementName();
 
     Navigator& m_navigator;
+    IsLoggedIn m_loginStatus;
```
I think it would be best to remove this and only keep this state in the network process.

pooneh-nb replied:
Initially we need to know what's the status of login. So we get the status as a string parameter:
https://developer.mozilla.org/en-US/docs/Web/API/NavigatorLogin/setStatus

charliewolfe replied:
But do we need to store it on this object? Right now it looks like we just pass it to the network process to be stored, so I don't understand why this value needs to be stored here.

pooneh-nb replied:
I need at least a variable to keep the status when deciding about whether it is 'logged-in' or 'logged-out'. Does that work if I define it before he if else statement?
```
IsLoggedIn loginStatus;
if (status == WTF::String::fromLatin1("logged-in"))
    loginStatus = IsLoggedIn::Yes;
else if (status == WTF::String::fromLatin1("logged-out"))
    loginStatus = IsLoggedIn::No;
else {
    promise->reject();
    return;
}
```

charliewolfe replied:
Yes, a local variable should be fine.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,32 +58,43 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
-    promise->resolve();
+    Document* document = m_navigator.document();
+    RefPtr frame = document->frame();
+    if (!frame || !frame->isMainFrame()) {
```
We probably don't want this isMainFrame check.
```suggestion
    if (!frame) {
```

pooneh-nb replied:
Do we let iframes to set login status?

charliewolfe replied:
I'm not aware of anything that says we shouldn't, especially for same-origin iframes. I think we should remove this mainframe check for now. We can add this back later if we need it along with tests that verify the correct behavior within subframes.

charliewolfe replied:
Now we can remove this main frame check, since we have the `hasSameOrigin()` check above.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -45,13 +45,15 @@
 #include <WebCore/SQLiteStatement.h>
 #include <wtf/CallbackAggregator.h>
 #include <wtf/CrossThreadCopier.h>
+#include <wtf/Forward.h>
```
This should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -473,6 +475,17 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
+{
+    ASSERT(RunLoop::isMain());
+
+    // if (status == "logged-in"_s)
```
This should be removed.

pooneh-nb replied:
Didn't we use this to make sure we are not store in a DB while we're on the main thread?

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -111,6 +115,8 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCountedAndCanMa
     using StorageAccessPromptWasShown = WebCore::StorageAccessPromptWasShown;
     using StorageAccessScope = WebCore::StorageAccessScope;
     using RequestStorageAccessResult = WebCore::RequestStorageAccessResult;
+    using Domain = WebCore::RegistrableDomain;
```
This is a bit weird. I think we could remove this and just use `RegistrableDomain` in the `setLoginStatus` function declaration.

pooneh-nb replied:
I removed it, but I don't remember if I added this line

pooneh-nb replied:
apparently removing it causes error in this:
```
void NetworkConnectionToWebProcess::setLoginStatus(...) {
resourceLoadStatistics->setLoginStatus(WTFMove(**domain**), loggedInStatus, WTFMove(completionHandler));}
```

charliewolfe replied:
Hmm I wouldn't expect removing this to cause a compiler error there. Make sure you also update the `setLoginStatus` function declaration below in this file to use `RegistrableDomain&&` instead of `Domain&&`.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -258,6 +265,7 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCountedAndCanMa
 
     HashSet<RegistrableDomain> m_domainsWithUserInteractionQuirk;
     HashMap<TopFrameDomain, Vector<SubResourceDomain>> m_domainsWithCrossPageStorageAccessQuirk;
+    HashMap<WebCore::RegistrableDomain, IsLoggedIn> m_loginStatus;
```
```suggestion
    HashMap<RegistrableDomain, IsLoggedIn> m_loginStatus;
```

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -60,6 +60,7 @@
 #include "RTCDataChannelRemoteManagerProxy.h"
 #include "RemoteWorkerType.h"
 #include "ServiceWorkerFetchTaskMessages.h"
+// #include "WebCore/LoginStatus.h"
```
This should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -91,8 +92,10 @@
 #include <WebCore/SameSiteInfo.h>
 #include <WebCore/SecurityPolicy.h>
 #include <optional>
+#include <wtf/Forward.h>
```
This should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -91,8 +92,10 @@
+#include <wtf/Forward.h>
 #include <wtf/HashSet.h>
 #include <wtf/LogInitialization.h>
+#include <wtf/StdLibExtras.h>
```
This should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1154,6 +1157,20 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
+}
+
+
+
```
Some of these new lines should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1167,6 +1184,7 @@ void NetworkConnectionToWebProcess::requestStorageAccessUnderOpener(WebCore::Reg
     }
 }
 
+
```
This new line should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -35,6 +35,7 @@
 #include "NetworkResourceLoadMap.h"
 #include "PolicyDecision.h"
 #include "SandboxExtension.h"
+// #include "WebCore/LoginStatus.h"
```
This should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -56,11 +57,13 @@
 #include <WebCore/RegistrableDomain.h>
 #include <WebCore/WebSocketIdentifier.h>
 #include <optional>
+#include <wtf/Forward.h>
```
This should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -56,11 +57,13 @@
 #include <wtf/RefCounted.h>
 #include <wtf/URLHash.h>
 
+
```
This should be removed.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -88,7 +92,7 @@ struct SameSiteInfo;
 enum class HTTPCookieAcceptPolicy : uint8_t;
 enum class IncludeSecureCookies : bool;
-}
+enum class IsLoggedIn : uint8_t; }
```
```suggestion
}
```

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -354,6 +358,8 @@ class NetworkConnectionToWebProcess
     void requestStorageAccessUnderOpener(WebCore::RegistrableDomain&& domainInNeedOfStorageAccess, WebCore::PageIdentifier openerPageID, WebCore::RegistrableDomain&& openerDomain);
 
+    void setLoginStatus(RegistrableDomain&&, WebCore::IsLoggedIn, CompletionHandler<void()>&&);
+    // void isLoggedIn(RegistrableDomain&& domain, CompletionHandler<void()>&&);
```
This should be removed

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1656,6 +1660,16 @@ void WebChromeClient::requestStorageAccess(RegistrableDomain&& subFrameDomain, R
+    protectedPage()->setLoginStatus(WTFMove(domain), loggedInStatus, WTFMove(completionHandler));
+}
+
+// void WebChromeClient::isLoggedIn(RegistrableDomain&& domain, CompletionHandler<void()>&& comcompletionHandlerple)
```
This should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -26,8 +26,11 @@
 
 #pragma once
 
+// #include "WebCore/LoginStatus.h"
```
This should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -26,8 +26,11 @@
+
 #include <WebCore/ChromeClient.h>
 #include <WebCore/HTMLVideoElement.h>
+#include <wtf/Forward.h>
```
This should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -433,6 +437,9 @@ class WebChromeClient final : public WebCore::ChromeClient {
     bool hasPageLevelStorageAccess(const WebCore::RegistrableDomain& topLevelDomain, const WebCore::RegistrableDomain& resourceDomain) const final;
 
+    void setLoginStatus(WebCore::RegistrableDomain&&, WebCore::IsLoggedIn, WTF::CompletionHandler<void()>&&) final;
+    // void isLoggedIn(WebCore::RegistrableDomain&& domain, WTF::CompletionHandler<void(bool)>&&) final;
```
This should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -37,6 +40,7 @@ enum class CookieConsentDecisionResult : uint8_t;
 enum class DidFilterLinkDecoration : bool;
 enum class StorageAccessPromptWasShown : bool;
 enum class StorageAccessWasGranted : uint8_t;
+enum class IsLoggedIn : uint8_t;
```
This should be in alphabetical order, below `DidFilterLinkDecoration`.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8346,6 +8346,16 @@ void WebPage::requestStorageAccess(RegistrableDomain&& subFrameDomain, Registrab
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetLoginStatus(WTFMove(domain), loggedInStatus), WTFMove(completionHandler));
+}
+
+// void WebPage::isLoggedIn(RegistrableDomain&& domain, CompletionHandler<void(bool)>&& completionHandler)
```
This should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -104,6 +105,9 @@
 
+#include <wtf/Forward.h>
+#include <wtf/text/WTFString.h>
+
```
These should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -52,6 +52,7 @@
 #include "SharedBufferReference.h"
 #include "UserData.h"
 #include "WebColorChooser.h"
+#include "WebCore/LoginStatus.h"
```
When including from another namespace the include should have `<>` instead of `""`.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -473,6 +475,17 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
     });
 }
 
+void WebResourceLoadStatisticsStore::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
```
The end of this function should call the completion handler.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -26,11 +26,20 @@
 #include "Navigator.h"
+#include "Page.h"
+#include "RegistrableDomain.h"
+#include <wtf/Assertions.h>
```
We should remove this assertions include.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -26,11 +26,20 @@
 
 namespace WebCore {
 
+enum class IsLoggedIn : uint8_t;
```
We can remove this forward declaration, because we already included `LoginStatus.h`.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -29,6 +29,8 @@
 #include "Encoder.h"
 #include "StorageAccessStatus.h"
+#include "WebCore/NavigationRequester.h"
+#include "WebCore/RegistrableDomain.h"
```
These two includes don't look like they are needed.

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -41,6 +43,7 @@
 #include <wtf/Condition.h>
 #include <wtf/CrossThreadCopier.h>
 #include <wtf/Forward.h>
+#include <wtf/HashMap.h>
```
This one can be removed too.

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -56,10 +56,6 @@
-#include <wtf/HashCountedSet.h>
-#include <wtf/OptionSet.h>
-#include <wtf/RefCounted.h>
-#include <wtf/URLHash.h>
```
I don't think these includes should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -104,6 +104,9 @@
 #include <WebCore/WindowFeatures.h>
 
+#include <wtf/Forward.h>
+#include <wtf/text/WTFString.h>
```
These two includes should be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
```diff
@@ -28,13 +28,13 @@
 
 #include <WebCore/ChromeClient.h>
 #include <WebCore/HTMLVideoElement.h>
-#include <wtf/WeakRef.h>
```
This include probably shouldn't be removed.

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8346,6 +8346,11 @@ void WebPage::requestStorageAccess(RegistrableDomain&& subFrameDomain, Registrab
     });
 }
 
+void WebPage::setLoginStatus(RegistrableDomain&& domain, WebCore::IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
```
```suggestion
void WebPage::setLoginStatus(RegistrableDomain&& domain, IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
```

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -8346,6 +8346,11 @@ void WebPage::requestStorageAccess(RegistrableDomain&& subFrameDomain, Registrab
 
+void WebPage::setLoginStatus(RegistrableDomain&& domain, WebCore::IsLoggedIn loggedInStatus, CompletionHandler<void()>&& completionHandler)
+{
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetLoginStatus(WTFMove(domain), loggedInStatus), WTFMove(completionHandler));
```
We should use a smart pointer here for `Connection`:
```suggestion
    WebProcess::singleton().ensureNetworkProcessConnection().protectedConnection()->sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetLoginStatus(WTFMove(domain), loggedInStatus), WTFMove(completionHandler));
```

pooneh-nb replied:
why do we use protectedConnection() instead of connection?

charliewolfe replied:
`protectedConnection()` returns a smart pointer instead of a reference, new code is supposed to use smart pointers when calling non-trivial member functions on an object (see the [safer cpp guidelines](https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-smart-pointers-for-object-lifetime-management)).

charliewolfe commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
```diff
@@ -139,4 +139,6 @@ messages -> NetworkConnectionToWebProcess LegacyReceiver {
 
     LoadImageForDecoding(WebCore::ResourceRequest request, WebKit::WebPageProxyIdentifier pageID, size_t maximumBytesFromNetwork) -> (std::variant<WebCore::ResourceError, Ref<WebCore::FragmentedSharedBuffer>> result)
+
+    SetLoginStatus(WebCore::RegistrableDomain domain, enum:uint8_t WebCore::IsLoggedIn loggedInStatus) -> ()
```
We should probably disable this IPC endpoint if the login status API is not enabled. For an example, see mentions of `[EnabledIf='isWebTransportEnabled()']` in this file which does this for web transport.

charliewolfe reviewed (commented):
We should consider also implementing `isLoggedIn()` in this PR so that we can write tests for this change.

cdumez commented on Source/WebCore/page/LoginStatus.h:
```diff
@@ -34,6 +34,11 @@
 
 namespace WebCore {
 
+enum class IsLoggedIn : uint8_t {
```
nit: Could be on 1 line:
```
enum class IsLoggedIn : uint8_t { No, Yes };
```

cdumez commented on Source/WebCore/page/Navigator+LoginStatus.idl:
```diff
@@ -28,7 +28,6 @@
 ] partial interface Navigator {
-    Promise<undefined> setLoggedIn();
-    Promise<undefined> setLoggedOut();
+    Promise<undefined> setLoginStatus(DOMString status);
```
Where is the spec for this?

I am surprised this is not a string enumeration in the spec:
- https://webidl.spec.whatwg.org/#idl-enums

charliewolfe replied:
Yes, I think this should take a string enumeration instead. It also looks like the function name was changed to just `setStatus`.

https://fedidcg.github.io/FedCM/#login-status-javascript

cdumez commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,40 +54,92 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
+        return;
+    }
+    page->chrome().client().setLoginStatus(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), loginStatus, [weakThis = WeakPtr { *this }, promise = WTFMove(promise)] {
+        if (!weakThis)
```
Why do we need this check?

pooneh-nb replied:
I believe it ensures that the object still exists when the callback is executed. If weakThis is null, it means the object has been deleted or is no longer valid, and the callback should not proceed. 
Also I saw it's been used in `DocumentStorageAccess::requestStorageAccess` which is close to what I'm doing in terms of functionality:
```
void DocumentStorageAccess::requestStorageAccess(Ref<DeferredPromise>&& promise)
{
  ...
  page->chrome().client().requestStorageAccess(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), RegistrableDomain::uncheckedCreateFromHost(document->topOrigin().host()), *frame, m_storageAccessScope, [this, weakThis = WeakPtr { *this }, promise = WTFMove(promise)] (RequestStorageAccessResult result) mutable {
          if (!weakThis)
              return;
              ....
```

charliewolfe replied:
In our case, we don't need to capture `weakThis` at all, since we are not accessing anything on `this` within the lambda, just resolving the promise.

cdumez commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,40 +54,92 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
+        return;
+    }
+    page->chrome().client().isLoggedIn(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), [weakThis = WeakPtr { *this }, promise = WTFMove(promise)] (bool isLoggedIn) {
+        if (!weakThis)
```
Why do we need this check?

pooneh-nb replied:
same as above comment. Please let me know if there is anything wrong with it.

cdumez commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -473,6 +474,26 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
+    ASSERT(RunLoop::isMain());
+
+    if (m_loginStatus.contains(domain))
+        return completionHandler(m_loginStatus.get(domain) == IsLoggedIn::Yes);
```
This is doing a double HashMap lookup (once in `contains()` above and then again in `get()` here). I recommend using HashMap::find() and use the iterator to avoid the double hash look up.

cdumez replied:
e.g.
```
auto it = m_loginStatus.find(domain);
completionHandler(it != m_loginStatus.end() && it->value == IsLoggedIn::Yes);
```

cdumez commented on Source/WebKit/Shared/NetworkProcessPreferencesForWebProcess.h:
```diff
@@ -30,7 +30,6 @@ namespace WebKit {
 struct NetworkProcessPreferencesForWebProcess {
     bool isWebTransportEnabled { false };
     bool usesSingleWebProcess { false };
-
```
Unrelated change, please drop from PR.

charliewolfe replied:
Looks like this change still needs to be dropped.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -494,7 +494,7 @@ void WebProcessProxy::initializePreferencesForNetworkProcess(const WebPreference
     m_preferencesForNetworkProcess = NetworkProcessPreferencesForWebProcess {
         preferences.getBoolValueForKey(WebPreferencesKey::webTransportEnabledKey()),
-        processPool().usesSingleWebProcess(),
+        processPool().usesSingleWebProcess()
```
Unrelated change, please drop from PR.

charliewolfe replied:
This one too.

charliewolfe commented on LayoutTests/http/tests/login-status/login-status-api-getter-setter-functions.html:
```diff
@@ -0,0 +1,39 @@
+            try {
+                await navigator.setLoginStatus('logged-in');
+                testPassed("Successfully set login status to 'logged-in'.");
+                const isLoggedIn = navigator.isLoggedIn();
```
This one seems like it should have `await`, like the other isLoggedIn() call

charliewolfe commented on Source/WebCore/page/Navigator+LoginStatus.idl:
```diff
@@ -28,7 +28,6 @@
-    Promise<undefined> setLoggedIn();
-    Promise<undefined> setLoggedOut();
+    Promise<undefined> setLoginStatus(DOMString status);
     Promise<boolean> isLoggedIn();
```
Similar here, it looks like this was changed since John added this. This should just be an attribute named `login`.

See https://fedidcg.github.io/FedCM/#login-status-javascript

charliewolfe replied:
Actually, let's change this function to match the spec in a future PR, since its implementation may change after moving to persistent storage.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,40 +54,92 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
+    }
+
+    RefPtr document = m_navigator.document();
+    if (!document && hasSameOrigin()) {
```
```suggestion
    if (!document || !hasSameOrigin()) {
```

charliewolfe commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -473,6 +474,26 @@ void WebResourceLoadStatisticsStore::requestStorageAccess(RegistrableDomain&& su
+    if (loggedInStatus == IsLoggedIn::Yes)
+        m_loginStatus.set(domain, loggedInStatus);
+    else
+        m_loginStatus.remove(domain);
```
Maybe we should replace the if/else with just `m_loginStatus.set(domain, loggedInStatus);`.

pooneh-nb replied:
But we don't want to assign logged-out to a domain, right? We just want to remove them from the hash map bc every user is logged-out by default.

charliewolfe replied:
It would be the same behavior, it would just save us a few lines of code. Though this hash map isn't a permanent solution, so it doesn't matter much. Up to you.

cdumez replied:
I think removing entries from the map when possible is cleaner. We wouldn't want a forever-growing hash map.

cdumez commented on Source/WebCore/page/NavigatorLoginStatus.h:
```diff
@@ -26,29 +26,34 @@
     }
-    static void setLoggedIn(Navigator&, Ref<DeferredPromise>&&);
-    static void setLoggedOut(Navigator&, Ref<DeferredPromise>&&);
+    static void setLoginStatus(Navigator&, const String& status, Ref<DeferredPromise>&&);
```
Now that you're using an enum on IDL side, `status` could be a `LoginStatus` enum class here instead of a `String`.

cdumez commented on Source/WebCore/page/Navigator+LoginStatus.idl:
```diff
@@ -23,12 +23,16 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+enum LoginStatus {
```
I think you either want to rename `IsLoggedIn` in C++ to `LoginStatus`, or use `[ImplementedAs= IsLoggedIn]` here in the IDL.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,40 +53,70 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
+        promise->reject();
+        return;
+    }
+    page->chrome().client().isLoggedIn(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), [weakThis = WeakPtr { *this }, promise = WTFMove(promise)] (bool isLoggedIn) {
```
We don't need to capture `weakThis` here, because it's not being accessed in the lambda.
```suggestion
    page->chrome().client().isLoggedIn(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), [promise = WTFMove(promise)] (bool isLoggedIn) {
```

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,40 +53,70 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
+        promise->reject();
+        return;
+    }
+    page->chrome().client().setLoginStatus(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), isLoggedIn, [weakThis = WeakPtr { *this }, promise = WTFMove(promise)] {
```
```suggestion
    page->chrome().client().setLoginStatus(RegistrableDomain::uncheckedCreateFromHost(document->securityOrigin().host()), isLoggedIn, [promise = WTFMove(promise)] {
```

charliewolfe reviewed (commented):
`LayoutTests/tests-options.json` was deleted somehow, that change should be undone before landing this.

The tests `fast/dom/navigator-detached-no-crash.html` and `http/tests/is-logged-in/available-in-secure-contexts.https.html` are failing. They are should be updated to reflect the name change of `setLoggedIn` -> `setStatus` on navigator.

You'll also need to add `WebCore/IsLoggedIn.h` to Headers.cmake to fix the cmake buidls.

charliewolfe commented on Source/WebCore/page/NavigatorLoginStatus.cpp:
```diff
@@ -47,40 +53,70 @@ ASCIILiteral NavigatorLoginStatus::supplementName()
-    promise->resolve();
+    RefPtr document = m_navigator.document();
+    if (!document || !hasSameOrigin()) {
+        ASSERT_NOT_REACHED();
```
I don't think there is a guarantee these assertions will not be hit, we should remove them.

Commit:
Define setLoginStatus
https://bugs.webkit.org/show_bug.cgi?id=276346

Reviewed by Charlie Wolfe.

Define setLoginStatus and add code to send the status to network process.
The status is kept in memory (in a hashmap) in WebResourceLoadStatisticsStore.
In the next step, we will keep this status persistently in a DB on disk.

* LayoutTests/http/tests/is-logged-in/available-in-secure-contexts.https.html:
* LayoutTests/http/tests/login-status/login-status-api-getter-setter-functions-expected.txt: Added.
* LayoutTests/http/tests/login-status/login-status-api-getter-setter-functions.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/fast/dom/navigator-detached-no-crash-expected.txt:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::setLoginStatus):
(WebCore::ChromeClient::isLoggedIn):
* Source/WebCore/page/IsLoggedIn.h: Copied from Source/WebCore/page/Navigator+LoginStatus.idl.
* Source/WebCore/page/IsLoggedIn.idl: Added.
* Source/WebCore/page/Navigator+LoginStatus.idl:
* Source/WebCore/page/NavigatorLoginStatus.cpp:
(WebCore::NavigatorLoginStatus::setStatus):
(WebCore::NavigatorLoginStatus::hasSameOrigin const):
(WebCore::NavigatorLoginStatus::isLoggedIn):
(WebCore::NavigatorLoginStatus::setLoggedIn): Deleted.
(WebCore::NavigatorLoginStatus::setLoggedOut): Deleted.
* Source/WebCore/page/NavigatorLoginStatus.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::setLoginStatus):
(WebKit::WebResourceLoadStatisticsStore::isLoggedIn):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setLoginStatus):
(WebKit::NetworkConnectionToWebProcess::isLoggedIn):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::setLoginStatus):
(WebKit::WebChromeClient::isLoggedIn):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLoginStatus):
(WebKit::WebPage::isLoggedIn):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/281308@main
